lib/puppet/util/character_encoding.rb

```diff
@@ -51,46 +57,39 @@ def convert_to_utf_8(string)
           # UTF-8. We can only guess, so if the string is valid UTF-8 as
           # currently constituted, we default to assuming the string originated
           # in UTF-8 and do not transcode it - we only set external encoding.
-          value_to_encode.force_encoding(Encoding::UTF_8)
-        elsif transcodable?(value_to_encode)
-          # If the string is not currently valid UTF-8 but it can be transcoded,
-          # we can guess this string was not originally unicode. We need it to
-          # be now, so transcode it to UTF-8. For strings with original
-          # encodings like SHIFT_JIS, this should be the final result.
-          value_to_encode.encode!(Encoding::UTF_8)
+          return string.force_encoding(Encoding::UTF_8)
+        elsif currently_valid
+          # If the string is not currently valid UTF-8 but it can be transcoded
+          # (it is valid in its current encoding), we can guess this string was
+          # not originally unicode. Transcode it to UTF-8. For strings with
+          # original encodings like SHIFT_JIS, this should be the final result.
+          return string.encode!(Encoding::UTF_8)
         else
-          # If the string is neither valid UTF-8 as-is nor is transcodable, fail
-          # at this point. It requires user remediation.
+          # If the string is neither valid UTF-8 as-is nor valid in its current
+          # encoding, fail. It requires user remediation.
           raise EncodingError
         end
       rescue EncodingError => detail
-        # catch both our own self-determined failure to transcode as well as any
-        # error on ruby's part, ie Encoding::InvalidByteSequenceError or
-        # Encoding::UndefinedConversionError.
-        raise Puppet::Error, _("#{detail.inspect}: #{value_to_encode.dump} is not valid UTF-8 and cannot be transcoded by Puppet.")
+        # Catch both our own self-determined failure to transcode as well as any
+        # error on ruby's part, ie Encoding::UndefinedConversionError on a
+        # failure to encode!.
+        raise Puppet::Error.new(_("%{error}: %{value} is not valid UTF-8 and cannot be transcoded by Puppet.") %
+          { error: detail.inspect, value: string.dump }, detail)
```

Comment:
I tried out doing a Puppet.warning and returning nil, and then benchmarked by doing 100000 lookups of a group with five regular ascii test members, one valid unicode member, and one invalid unicode member and...
wow. It turns out adding `Puppet.warning` on failure to encode is almost 3x slower than not doing a warning (shave a second off that if we don't parse the preceding EncodingError), and almost 2x slower than just raising `Puppet::Error` and rescuing it in `Etc`.
With Puppet.warning instead of rescue/raise:
```
       user     system      total        real
::Etc lookup group with 5 members 100000 times  0.980000   0.610000   1.590000 (  1.596053)
Puppet::Etc with 5 members 100000 times  5.080000   1.030000   6.110000 (  6.125440)
```
Without Puppet.warning (just return nil):
```
       user     system      total        real
::Etc lookup group with 5 members 100000 times  0.920000   0.700000   1.620000 (  1.634029)
Puppet::Etc with 5 members 100000 times  1.830000   0.720000   2.550000 (  2.560253)
```
With exceptions for flow control (current PR):
```
       user     system      total        real
::Etc lookup group with 5 members 100000 times  0.970000   0.590000   1.560000 (  1.558745)
Puppet::Etc with 5 members 100000 times  2.580000   0.880000   3.460000 (  3.478575)
```

I expect that's because these benchmarks just call `::Etc` and `Puppet::Etc` so in this PR the rescue means we never invoke any of the logging machinery, just create the superfluous exception objects. If we do Puppet.warning in `convert_to_utf_8!` that is always invoked since default log level is warning.

If I do `Puppet.debug` instead of `Puppet.warning` with log level at default (warning) the performance is roughly on par with raising the exceptions. Still a performance hit over not logging at all, but probably the best mix of user experience/performance? I have updated the PR on this assumption.
With Puppet.debug instead of rescue/raise:
```
::Etc lookup group with 5 members 100000 times  0.970000   0.670000   1.640000 (  1.642895)
Puppet::Etc with 5 members 100000 times  2.600000   0.870000   3.470000 (  3.470088)
```